### PR TITLE
add support for arm-unknown-linux-gnueabihf

### PR DESCRIPTION
**Pull request by gnzlbg**

### CODE REVIEW

**Comment by japaric** [comment]
@gnzlbg  The Dockerfile is missing.

**Comment by gnzlbg** [comment]
@japaric sorry about that, should be fixed.

**Comment by gnzlbg** [comment]
Moved this into #157 to launch a homu build.

**Comment by gnzlbg** [comment]
@japaric said:

> I just remembered:
>
> > apt-get install -y --no-install-recommends g++-arm-linux-gnueabihf
>
> This is wrong for arm-unknown-linux-gnueabihf because that targets supports both ARMv6 and ARMv7 but g++-arm-linux-gnueabihf targets ARMv7 thus with this combination the binary will contain ARMv7 instructions and will crash when executed on a ARMv6 device. You won't detect this problem when testing under QEMU because qemu-arm emulates an ARMv7 processor by default. But quite a few people, including me, have run into this runtime crash (SIGILL) when using the Ubuntu toolchain for cross compilation.
>
> TL;DR The C toolchain (gcc + libc) must be compiled from scratch to target ARMv6

and then also said:

> You can probably copy paste some of the Docker build stuff from rust-lang/rust. There they use crosstool-ng to build an ARMv6 toolchain.

So this is what I have tried to do by basically copying rust-lang/rust's [dist-armhf-linux](https://github.com/rust-lang/rust/blob/master/src/ci/docker/dist-armhf-linux/Dockerfile). 

However, without a way to test this locally, I am basically going in blind.

**Comment by japaric** [comment]
I ran locally and it errored with:

``` console
Step 4/10 : RUN bash /crosstool-ng.sh
 ---> Running in 2f4ad86e3f0d
+ url=http://crosstool-ng.org/download/crosstool-ng/crosstool-ng-1.22.0.tar.bz2
+ curl -f http://crosstool-ng.org/download/crosstool-ng/crosstool-ng-1.22.0.tar.bz2
+ tar xjf -
/crosstool-ng.sh: line 14: curl: command not found
```

I can probably take a closer look at this tomorrow. I can also give you "try rights" so you can do `@homu try` yourself but edit-test loops on Travis are rather slow.

**Comment by gnzlbg** [comment]
@japaric I've added curl to the list of packages to install but yeah, the edit-debug cycle over travis or homu is going to be very long.

**Comment by japaric** [comment]
Sorry, I won't have time to look into this at least until the weekend. I have given you "try right" though in case you want to try to get a bit further using travis.

**Comment by gnzlbg** [comment]
@homunkulus try

-- @japaric if this fails (which probably will do) I'll disable temporarily all other build bots and see how far I come before the weekend. It is not a pressing issue, but I'd really love to migrate the `stdsimd` crate to cross, in particular since @malbarbo is working towards allowing qemu-system to run which is something we need to properly test run-time feature detection (qemu-user doesn't emulate `/proc/cpuinfo`, `auxv`, etc.).

**Comment by homunkulus** [comment]
:hourglass: Trying commit 3115d90037e5a80a983b4a6918e8672007ab26a8 with merge 5ae8b2b47c7405a2ff0cf3df228c09a5b9f12290...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/302727010?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 46b39f804b1625a846740ee73dd0bbb12ca64b42 with merge 0188f1498db4109347805ec251d796ae93216c1b...

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/japaric/cross/builds/302897446?utm_source=github_status&utm_medium=notification)
State: approved= try=True

**Comment by gnzlbg** [comment]
Looks like this worked, doing a full run: @homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 24bd3de3fa2daac3e1a866b886c7342bb21697c9 with merge e1779e814c08d3b2ec20f7f64344fd6ff121efd3...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/302901726?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 7f7298b94ac5434d76f9045125016b0cd53c1221 with merge fec6640880e7ac436453f808240b0672a9f257b9...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/302927563?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 3a3d151303e58474dff94988d5cd7937dde96dc5 with merge b7b0761c1254b408591a8d345c90ddea6ab20fe7...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/302989743?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 872ff3b7a53028e6a470a7db4df04b890eea298a with merge ad2b0b47990935f584b294500c3859d6a6a9a392...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303026322?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 9c77ed801add5a852e265b2e4467aceb27befbd3 with merge a1096427ba692c9fed24bafc401919de73c969e7...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303073895?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 744d4ea7011e5ae3fcda90dd64108461f8e4d3e5 with merge 5b8a0d73546c6cecf1c3782a32f86fec855e23d5...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303076222?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 8ddd0baa7bafae91af118b0e8f38e5adfc8e88be with merge 4aaa8c4fdb50c670e3dbdf0e6379d59f09a9ebd7...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303080352?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit edc9a733b35f95c3df9884575b2193e08833aa80 with merge 40ef8aba6d805779c4f5b083d37a10424f5e2ea2...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303083342?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 96a81ebe41875782b24b119fadc21200dbea8d61 with merge 01815d8fcd4fd862f28dba29a8f5bf6de5fbab8e...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303086100?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit b2cea43f3a0b64ed388c2d87e0d02d84cf281513 with merge 26aaff2b31c1c913eef2896a03289a6246b19d6b...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303089391?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit b57a483ad2ac5f044ad6b6eeab392252593dbd11 with merge 6dc815553c59d647e4db9e6d9fa1be28f3b497d1...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303091658?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 8aa2023220d2b90d9d2fbba66257a4f1050c75eb with merge 6447a019e014ba0f4fbb1b607d252db44e9dd3a9...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303095590?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 6cb7729d26acb19abfe6807f1b6b678f356dbc88 with merge b2c904a9d6f415c4f9766e090878d8ee1a26fb5c...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303100819?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit e92b720cd1e9e9d2db4da5e5937e1cc32eaaa270 with merge 30090c93ffc1ce09c87cd6625e381fd8041a346a...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303109499?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 74524541f438b62c16df579ba05888846efa0f03 with merge 1607a1205998b24d61d86164fc9930b06929438e...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303217859?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 9d215a4b092548e29ac66f5c955dd049e7d70159 with merge d09f9f023fa02b3856c095aa8ebcb5e380ab5bac...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303446001?utm_source=github_status&utm_medium=notification)

**Comment by homunkulus** [comment]
:umbrella: The latest upstream changes (presumably #157) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by gnzlbg** [comment]
@japaric it seems it takes too long for this to build

**Comment by homunkulus** [comment]
:umbrella: The latest upstream changes (presumably 94ab2acdf99af542374151b1c20fb39fa1491bc9) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by homunkulus** [comment]
:umbrella: The latest upstream changes (presumably #170) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by gnzlbg** [comment]
@japaric any chance of looking into this?

**Comment by malbarbo** [comment]
The build seems to be taking too long. How about using a pre-build toolchain from https://toolchains.free-electrons.com?

**Comment by Dylan-DPC** [comment]
bors: r+

**Comment by Dylan-DPC** [comment]
@gnzlbg can you merge these conflicts if possible? Looking to merge this after that.

**Comment by gnzlbg** [comment]
This never passed tests because of what @malbarbo mentioned here: https://github.com/rust-embedded/cross/pull/158#issuecomment-360763050

**Comment by Dylan-DPC** [comment]
Yeah but we resolved that issue in the other PRs which we already merged.

**Comment by gnzlbg** [comment]
So i've resolved the conflicts here, but I don't think this will work as is.

**Comment by gnzlbg** [comment]
If someone wants to pick it up the best would probably be to just fetch my branch and submit a new PR.

**Comment by Dylan-DPC** [comment]
What's missing right now? If you can list it, it will be easier for someone to pick it up

**Comment by Dylan-DPC** [comment]
@legokichi great. Are you open to making a PR/pushing a commit to this PR to make the relevant changes?

**Comment by legokichi** [comment]
i have no time

**Comment by reitermarkus** [comment]
This can probably be closed now that https://github.com/rust-embedded/cross/pull/245 has been merged.

**Comment by Dylan-DPC** [comment]
closing this in favour of #245. If anything wasn't added in that PR, we can make a new PR out of it.
